File: code/ch12/12.4.2.GetMoneyAmount.cpp
class Solution {
private:
    int cost(int low, int high, vector<vector<int>> mem) {
        if (low >= high)
            return 0;
        
        if (mem[low][high])
            return mem[low][high];
        
        int res = INT_MAX;
        for (int i = (low + high) / 2; i < high + 1; i++) {
            int tmp = i + max(cost(low, i - 1, mem), cost(i + 1, high, mem));
            res = min(res, tmp);
        }
        mem[low][high] = res;
        
        return res;
    }
public:
    int getMoneyAmount(int n) {
        if (n == 1) return 0;

        vector<vector<int>> mem(n + 1, vector<int>(n + 1, 0));
        return cost(1, n, mem);
    }
};

File: code/ch12/12.4.2.GetMoneyAmount.py
import sys
from typing import List


class Solution:
    def getMoneyAmount(self, n: int) -> int:
        def cost(low: int, high: int, mem: List[List[int]]) -> int:

            if low >= high:
                return 0
            if mem[low][high] != 0:
                return mem[low][high]

            res = sys.maxsize
            for i in range((low + high) // 2, high + 1):
                tmp = i + max(
                    cost(low, i - 1, mem),
                    cost(i + 1, high, mem),
                )
                res = min(res, tmp)
            mem[low][high] = res
            return res

        if n == 1:
            return 0
        mem = [[0] * (n + 1) for _ in range(n + 1)]

        return cost(1, n, mem)
